Followers: First pass at blocking accounts

obenland opened the pull request:
![image](https://github.com/user-attachments/assets/426412a5-31d2-4be2-8f26-a1821ce92e10)

![Screenshot 2025-07-15 at 3 13 20 PM](https://github.com/user-attachments/assets/ad154134-44c5-4a5e-9389-bf8ff8d40c8a)

Depends on #2027

## Proposed changes:
This PR proposes a two-step block process, that allows users to block accounts individually or in bulk. It comes with an "Are you sure?" step, similar to how User deletes work in Core.

This is probably something we might want to add to the Following list as well, in which case we probably should think about a parent list class that follower and following can inherit from.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?

## Testing instructions:

* Go to Users > Followers ⁂ in wp-admin
* Click "Block" link next to any follower OR select multiple followers and choose "Block" from bulk actions
* Verify confirmation screen shows correct account details and blocking options
* Test blocking with both individual and site-wide options checked
* Confirm blocked actors appear in Users > Blocked Actors ⁂ page
* Test unblocking functionality from the Blocked Actors page
* Verify appropriate permissions (only users with 'manage_options' can do site-wide blocks)

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [x] Added - for new features
-   [ ] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

Add follower blocking functionality with confirmation screens and Blocked Actors management

</details>

obenland committed:
Followers: First pass at blocking accounts

obenland committed:
Let's make CLI changes separate

obenland committed:
DRY it up

Jiwoon-Kim commented:
Hmm… I think this could become a more general-purpose implementation if blocking were possible not only between remote actors but also between local actors. While this may not be very meaningful for personal WordPress use cases (as opposed to forums or social networks), it could help maintain logical consistency.

Also, since blog actors are not the instance themselves, I believe we should distinguish between object types when handling blocking. In that sense, separating the ability to block individual users and to block at the instance level (remote user/instance) seems like a more objective approach.

Additionally, at the instance level, it might be more useful to allow for moderation-level actions such as soft-hiding or de-prioritizing content rather than enforcing a strict block. This would provide more nuanced control over federation behavior, especially in cases where full blocking may be too aggressive or unnecessary.

obenland committed:
Account for multiple users

obenland committed:
Add changelog

Follower lists now include the option to block individual accounts, providing users with more control over their interactions.

pfefferle commented on includes/table/class-followers.php:
```diff
@@ -104,6 +104,90 @@ public function process_action() {
+
+				// Handle single follower block.
+				if ( isset( $_GET['follower'], $_GET['_wpnonce'] ) ) {
+					$follower = \esc_url_raw( \wp_unslash( $_GET['follower'] ) );
```
why not also use the ID? it makes the code easier to maintain and allow a more generic handling of the block mechanism below.

obenland replied:
Fixed in [dc4ecc7](https://github.com/Automattic/wordpress-activitypub/pull/1935/commits/dc4ecc7e838e0fa8b89410db6dd57d650b04449b).

pfefferle commented on includes/table/class-followers.php:
```diff
@@ -104,6 +104,90 @@ public function process_action() {
+							// If user is admin for blog actor, also remove relationships for blog actor.
+							if ( \user_can( $this->user_id, 'manage_options' ) ) {
+								Follower_Collection::remove( $follower, Actors::BLOG_USER_ID );
+							}
```
what about encapsulating that part, to be able to use it for single and bulk handling?

obenland committed:
Update to using post ID

pfefferle commented:
I gave that case some more thought. it's a tricky one because it mixes up user-specific actions with blog-wide behavior.

Scenario:

* User A blocks a profile. As a result, that user is removed from their personal followers list, but also added to the global block list.
* Now, if User B also had that same remote user as a follower, they might be confused when they suddenly stop receiving messages from them.

Jiwoon-Kim commented:
I think it would be helpful to separate user-level block lists (similar to following/follower fields) from instance-level moderation tools. For example, soft-hiding or excluding flagged accounts from search results. It’s not clear how blocking is currently implemented, but UI-based filtering using per-user blocklists might be necessary.

I originally thought implementing a site-level superuser role would help manage this better — not for enforcing instance-wide blocks, but to allow individual users to block or unblock others on their own terms. This would enable more nuanced control without resorting to blanket moderation.

obenland commented:
> I gave that case some more thought. it's a tricky one because it mixes up user-specific actions with blog-wide behavior.

Agreed, that's generally the case with how blocking currently works.

We could limit the inline action to the site owner, to make it essentially an instance-level tool?

pfefferle commented:
> We could limit the inline action to the site owner, to make it essentially an instance-level tool?

might be a good compromise to start with. we could limit the block action to the blog-user UI, but we should be consistent with the follow up process. after a block, we should:

* remove all connections to all local users, so that no users will see the remote-actor in a local followers/following list.
* if someone follows the blockerd user, we should maybe also send an email notification, that the site admin blocked this account!?

obenland committed:
Refactor follower blocking logic into block_followers method

Extracted and centralized the logic for blocking one or more followers into a new private block_followers() method. Updated single and bulk block actions to use this method, improving code reuse and maintainability. Enhanced confirmation screens to display usernames and improved error handling and messaging.

pfefferle commented:
we should also block outgoing follow requests for accounts on the block list!?

pfefferle commented:
So maybe it makes sense to once again evaluate a dedicated block list aside the comment block list.

Jiwoon-Kim commented:
I realize this is a separate issue, but just to mention — I personally dislike email notifications since I rarely check my inbox. I think a PWA-integrated notification block would be much more useful, though I understand it's currently a low priority.

Jiwoon-Kim commented:
Mute(user) / Silence(admin)

1. https://misskey.io/settings/mute-block
2. https://mastodon.social/severed_relationships
 `https://mastodon.social/severed_relationships`

The `https://mastodon.social/severed_relationships` page lists **publicly severed federation relationships** between the Mastodon instance and other remote instances.

## 🔍 What It Shows

Each entry on the page typically includes:

- **Domain Name**: The remote instance that is blocked or defederated.
- **Reason (optional)**: Such as spam, abuse, hate speech, or user safety.
- **Block Type**:
  - `suspend`: Full block, removes all content and blocks all federation.
  - `silence`: Federation remains, but content does not appear in public timelines.
  - `limit`: Partial feature restrictions.
- **Date of Action**: When the block or limitation was enforced.

## 🎯 Purpose

This list serves to maintain **transparency** and support **community-level moderation** in the decentralized Fediverse environment. Other instance administrators can:

- Reference it when forming their own moderation policies.
- Understand the context of federation behavior with Mastodon.social.

## 🪟 Relevance to WordPress + ActivityPub

If you're using a federated WordPress site via the ActivityPub plugin, a similar feature could enhance moderation capabilities:

- Ability to **block or silence remote actors or instances** at the site level.
- Maintain a **public block list** similar to `severed_relationships`.
- Support nuanced moderation strategies like:
  - Soft-hiding actors
  - Deprioritizing content in feeds or search results
- Provide **transparency and discoverability** regarding federation policies.

## ✅ Suggestions for WordPress ActivityPub Plugin

- Introduce a `severed_relationships`-style endpoint (e.g., `/severed_relationships.json`)  
- Add admin panel tools for managing:
  - Instance-level blocks
  - Actor-level blocks
  - Block reason and metadata
- Make moderation actions visible in a public or semi-public way, as an opt-in feature

pfefferle commented:
@obenland there is a merge conflict and what do you think about my latest comment:

> So maybe it makes sense to once again evaluate a dedicated block list aside the comment block list.

(I am not ignoring this PR)

obenland commented:
Yeah, based on that comment I've deprioritized this a bit. Can pick it up again later today maybe or next week. I switched the PR back to draft to better reflect its actual status

obenland committed:
Update block action to use Moderation methods instead of disallowed_keys

Replace the legacy disallowed_keys approach with the new Moderation class methods:
- Use Moderation::add_user_block() for user-specific blocks
- Use Moderation::add_site_block() for site-wide blocks when user has admin permissions
- Update confirmation dialog text to reflect the new moderation system
- Improve error handling to only remove followers if blocking succeeds

obenland committed:
Refactor follower removal to use hook-based approach

- Add remove_blocked_actors method to Followers collection
- Register hook to automatically remove followers when actors are blocked
- Use Actors::get_id_by_various for better performance (only gets ID, not full post)
- Simplify block_followers method to rely on hooks for follower removal
- Maintain consistency with existing Following hook pattern

obenland committed:
Move block confirmation screens to templates and add site-wide blocking option

- Create block-confirmation.php template for single account blocking
- Create bulk-block-confirmation.php template for bulk account blocking
- Refactor show_block_confirmation() and show_bulk_block_confirmation() to use templates
- Add optional site-wide blocking checkbox for admin users
- Use submit_button() for proper WordPress admin styling
- Update blocking logic to only add site-wide blocks when explicitly requested
- Improve separation of concerns between logic and presentation

This gives admins explicit control over personal vs site-wide blocking and follows
WordPress templating best practices.

obenland committed:
Refactor confirmation screens to be completely template-based

- Move all logic from show_block_confirmation() into block-confirmation.php template
- Move all logic from show_bulk_block_confirmation() into bulk-block-confirmation.php template
- Update process_action() to call templates directly with minimal arguments
- Remove now-unused show_block_confirmation() and show_bulk_block_confirmation() methods
- Templates are now self-contained and handle their own validation and data preparation

This eliminates ~70 lines of code from the Followers class and creates a cleaner
separation between the table logic and confirmation screen presentation.

obenland committed:
Align variable assignment formatting in templates

Standardizes the spacing and alignment of variable assignments in class-followers.php, block-confirmation.php, and bulk-block-confirmation.php for improved code readability and consistency.

obenland committed:
Merge branch 'feature/new-actor-blocks' into add/block-action

obenland committed:
Fix redirect and update block confirmation text

Adds 'confirm' to the list of query args removed on block action redirect in class-followers.php. Also updates the block confirmation message in both block-confirmation.php and bulk-block-confirmation.php to end with a period for consistency.

obenland committed:
Prevent following blocked accounts in admin table

Adds a check in the Following admin table to prevent users from following accounts they have blocked. Displays an error message if a follow attempt is made on a blocked account.

obenland committed:
Merge branch 'feature/new-actor-blocks' into add/block-action

obenland committed:
Merge branch 'feature/new-actor-blocks' into add/block-action

obenland committed:
Add site-wide block check when following accounts

Introduces a check to prevent following accounts that are blocked site-wide, in addition to user-specific blocks. Displays an appropriate error message and updates the redirect URL when a site-wide block is detected.

obenland committed:
Merge branch 'feature/new-actor-blocks' into add/block-action

obenland committed:
Refactor block confirmation templates for clarity and permissions

Replaces usage of user_can with current_user_can for permission checks, removes unnecessary variables, and improves code clarity in block-confirmation.php and bulk-block-confirmation.php. Also adjusts placement of informational text and streamlines nonce field generation.

obenland committed:
Fix unblock reference text in block confirmation templates

Update text to reference 'Blocked Actors list' instead of 'ActivityPub moderation settings' to accurately reflect where users can unblock accounts.

pfefferle commented on includes/wp-admin/table/class-following.php:
```diff
@@ -123,11 +124,34 @@ public function process_action() {
+				}
+
+				// Check if site has blocked the account.
+				if ( \in_array( $profile, Moderation::get_site_blocks()['actors'], true ) ) {
```
we should have a dedicated function for that check, something like `Moderation::is_actor_blocked( $uri )` to not have too much array comparisons in the places we want to have this check.

pfefferle replied:
This way, we could also easily support any kind of Actor representation (Object, Array or URI).

obenland replied:
Sure! Okay to do in a follow-up?

pfefferle commented on templates/bulk-block-confirmation.php:
```diff
@@ -0,0 +1,88 @@
+		<?php endif; ?>
+
+		<p class="submit">
+			<?php submit_button( __( 'Confirm Block', 'activitypub' ), 'primary', 'submit', false ); ?>
```
KISS.

```suggestion
			<?php submit_button( __( 'Confirm', 'activitypub' ), 'primary', 'submit', false ); ?>
```

pfefferle commented on templates/block-confirmation.php:
```diff
@@ -0,0 +1,71 @@
+		<?php endif; ?>
+
+		<p class="submit">
+			<?php submit_button( __( 'Confirm Block', 'activitypub' ), 'primary', 'submit', false ); ?>
```
```suggestion
			<?php submit_button( __( 'Confirm', 'activitypub' ), 'primary', 'submit', false ); ?>
```

obenland committed:
Update templates/block-confirmation.php

Co-authored-by: Matthias Pfefferle <pfefferle@users.noreply.github.com>

obenland committed:
Update templates/bulk-block-confirmation.php

Co-authored-by: Matthias Pfefferle <pfefferle@users.noreply.github.com>

obenland committed:
Merge branch 'feature/new-actor-blocks' into add/block-action

pfefferle commented:
let's merge this one into the feature branch. it feels easier to review the PR this way.
